tuf/formats.py

```diff
@@ -793,7 +793,13 @@ def make_fileinfo(length, hashes, version=None, custom=None):
     information of a metadata or target file.
   """
 
-  fileinfo = {'length' : length, 'hashes' : hashes}
+  fileinfo = {}
+
+  if length is not None:
```

Comment:
I think we could just use the "truthey" check here (`if length:`), and in the following IF checks, which excludes all "falsey" values at least

Reply:
Is it valid to have zero-length files in a repository? For length especially it feels safer to explicitly check for `None`?

Reply:
I feel like this line can be simplified to `length = length or None`, because we really have too many schema checks and such all over the place, which is not necessary for writing metadata (makes code clunky, especially in a dynamically typed language) so much as reading them.

Reply:
Agreed, `fileinfo['length'] = length or None` is concise and readable.

Reply:
Yea, I agree about that `fileinfo['length'] = length or None` is a good approach.
Maybe it would be good to change the checks for the other parameters and make them one-liners as well?

Reply:
I just realized that by using `fileinfo['length'] = length or None` there is an unexpected side effect.
If I use the proposed one-liner to check all arguments and I call `make_fileinfo` with all arguments with a value of `None` then I will end up with the dictionary:
`fileinfo = {"length": None, "hashes": None, "version": None, "custom": None}`

but if I use the if statements and call the function the same way I will end up with just:
`fileinfo = {}`.

What of the above effects do we want?

Reply:
And I noticed another problem.
If I decide to use this one-liner `fileinfo['length'] = length or None` only for length and run the tests I encounter this problem in 6 of the tests:
![image](https://user-images.githubusercontent.com/16246778/82338363-7c51e300-99f5-11ea-82cc-27704f9dfc3e.png)

The problem is caused by the same fact I pointed out in the previous comment. 
At the end of the `make_fileinfo` there is this check `FILEINFO_SCHEMA.check_match(fileinfo)`.
File info dictionary during this check is `fileinfo = {"length": None, "hashes": X, "version": Y, "custom": Z}.

In `check_match` function `fileinfo["hashes"]` is found and is checked again recursively by the same function:
![image](https://user-images.githubusercontent.com/16246778/82339013-3b0e0300-99f6-11ea-8df7-952428e4db63.png)

but because the value is none, then the above error is thrown.

Reply:
Ah, right. I should have looked at the context for this change/discussion more. Apologies.

I would expect that `make_fileinfo()` becomes `make_targets_fileinfo()` and remains as currently implemented (only changing `FILEINFO_SCHEMA` -> `TARGETS_FILEINFO_SCHEMA`, per the discussion above).

You could add a new helper, `make_metadata_fileinfo(version, length=None, hashes=None)` and then do as Trishank suggested above and check for truthy values, i.e.:
```python
if hashes:
  fileinfo['hashes'] = hashes
```

Reply:
@joshuagl I addressed your comments in a new commit.